Spring Security Bonus Module (#54)
LaunchCodeEducation/java-web-dev-curriculum: content/authentication/next-steps/bonus-module/user-data/_index.md

---
title: "User Owned Data"
date: 2023-11-14T09:28:27-05:00
draft: false
weight: 1
originalAuthor: Ben Clark # to be set by page creator
originalAuthorGitHub: brclark # to be set by page creator
reviewer: <no-value> # to be set by the page reviewer
reviewerGitHub: <no-value> # to be set by the page reviewer
lastEditor: # update any time edits are made after review
lastEditorGitHub: # update any time edits are made after review
lastMod: # UPDATE ANY TIME CHANGES ARE MADE
---

With the authentication filter additions, our application now requires users
to log in before they can access any features of the site. However, users
will want the data that they create to be associated with their account. How
can we make sure that the events and categories that a user create will be

> associated with their account?

Reviewer: ```suggestion
will want the data that they create to be associated with their account. How
can we ensure that the events and categories a user creates are
associated with their account?
```

> Let's associate the new data with the logged-in user using one-to-many database relationships.

Reviewer: ```suggestion
Let's use a one-to-many relationship to associate the new data with the logged-in user.
```

Author: I rewrote the intro to this section to reduce the redundancy of explanation

We will specify that a `User` has a one-to-many relationship with the `Event` objects
that they create, along with `EventCategory` and `Tag`. This way, we can specify that
we want to load all events or save an event *for a specific user*.

{{% notice blue Note "rocket" %}}
The code for this section begins with the [auth-filter branch](https://github.com/LaunchCodeEducation/CodingEventsJava/tree/auth-filter) and ends with the [user-data branch](https://github.com/LaunchCodeEducation/CodingEventsJava/tree/user-data) of the `CodingEventsJava` repository.
{{% /notice %}}

## Creating User Specific Data - Video

**TODO**

## Creating User Specific Data - Text

### Updating the models with a `User` field

Users *own* their data when the entities that they create (events, categories, etc) have their
`user_id` associated with each new entity as a foreign key. That would allow us to, 
say "get all events for a specific user". We need to set up a **One-To-Many** relationship

> between the `User` model and the different data models (`Event`, `EventCategory`, `Tag`).

Reviewer: ```suggestion
To set up a one-to-many relationship between the `User` model and the other data models (`Event`, `EventCategory`, and `Tag`), we need to add `user_id` to the other data models as a foreign key. 
```

Reviewer: You already said that we need to set up a one-to-many relationship above so the final sentence of this paragraph is redundant.

Open the `Event` model and add a `User` field to the field definitions. We will call it `creator`
as it's a reference to the user that created the event, and we will give it a `@ManyToOne` annotation

> as there can be many events associated with a single creator.

Reviewer: ```suggestion
Open the `Event` model and add a `User` field to the field definitions. Call it `creator` and give it a `@ManyToOne` annotation.
```

```java
    @ManyToOne
    private User creator;
```

We'll also need to add a getter and setter. Add the following to your getters

> and setters in `Event` class.

Reviewer: ```suggestion
We'll also need to add a getter and setter for `creator`.
```

```java
    public User getCreator() {
        return creator;
    }

    public void setCreator(User creator) {
        this.creator = creator;
    }
```

`Event` is not the only model that we want to be user-owned. **Repeat the above

> steps** to add the `creator` field/getters/setters to `EventCategory`.

Reviewer: ```suggestion
Repeat the above steps to add the `creator` field/getters/setters to `EventCategory`.
```

{{% notice blue Note "rocket" %}}
The `Tag` resource will be your task to update throughout this bonus module. We
will leave hints that you should continue updating that resource but will not
explicitly include instructions on keeping the tag feature updated.

You should update the `Tag` class as well to track `User creator`.
{{% /notice %}}

### Saving the `User` when creating new data

> We can store an associated User as the creator of an Event/Category/Tag. Next

Reviewer: ```suggestion
We can store an associated `User` as the creator of an `Event`/`EventCategory`/`Tag`. Next
```

> we need to make sure that the currently logged-in User is set as the creator

Reviewer: ```suggestion
we need to make sure that the currently logged-in user is set as the creator
```

before saving new entries. Let's update the `EventController` first to set

> the `User creator` field in new events.

Reviewer: ```suggestion
the `creator` field in new events.
```

of passing in all categories. Update your function like below.

```java{hl_lines="2-3 6"}
    @GetMapping("create")
    public String displayCreateEventForm(Model model, HttpSession session) {
        User currUser = authController.getUserFromSession(session);
        model.addAttribute("title", "Create Event");
        model.addAttribute(new Event());
        model.addAttribute("categories", eventCategoryRepository.findAllByCreator(currUser));
        return "events/create";
    }
```

Don't forget to update the same in the **error** case of the POST request handler. If
there are form errors, we want to pass the user-created categories back to the form.

```java{hl_lines="7"}
    @PostMapping("create")
    public String processCreateEventForm(@ModelAttribute @Valid Event newEvent,
                                         Errors errors, Model model, HttpSession session) {
        User currUser = authController.getUserFromSession(session);
        if(errors.hasErrors()) {
            model.addAttribute("title", "Create Event");
            model.addAttribute("categories", eventCategoryRepository.findAllByCreator(currUser));
            return "events/create";

>         }

Reviewer: Is this method missing a closing bracket?

```

When we display the delete events form, we want to make sure it displays the user-created
events. Let's repeat the same procress to retrieve events for the current user in the
`displayDeleteEventsForm` method.

```java{hl_lines="2-3 5"}
    @GetMapping("delete")
    public String displayDeleteEventForm(Model model, HttpSession session) {
        User currUser = authController.getUserFromSession(session);
        model.addAttribute("title", "Delete Events");
        model.addAttribute("events", eventRepository.findAllByCreator(currUser));
        return "events/delete";
    }
```

Finally, we want to make sure that the event details page will show events owned by the
currently logged in user and reject any event ID's owned by other users. Once again,
we'll use `HttpSession` and `AuthenticationController` to retrieve the current user,
and we'll retrieve the event based on its ID and the current user.

In the `displayEventDetails` method, let's add:

```java{hl_lines="2-5"}
    @GetMapping("detail")
    public String displayEventDetails(@RequestParam Integer eventId, Model model, HttpSession session) {
        User currUser = authController.getUserFromSession(session);

>         Optional<Event> result = eventRepository.findByIdAndCreator(eventId, currUser);

Reviewer: If you are only showing part of a method, I would add some contextual clues so they know for sure where to add the code. See the suggestion below for how I would do it.

Reviewer: ```suggestion
    @GetMapping("detail")
    public String displayEventDetails(@RequestParam Integer eventId, Model model, HttpSession session) {
        User currUser = authController.getUserFromSession(session);

        Optional<Event> result = eventRepository.findByIdAndCreator(eventId, currUser);
        
        // Rest of our method
        }
```
